File: typescript/src/middleware/middleware.ts
import type { CallerContext } from "./types.js";
import type { PolicyEvaluator } from "./policy.js";
import type { Auditor } from "./audit.js";
import { Action } from "./types.js";

export interface ToolExecutor {
  execute(toolName: string, args: Record<string, unknown>): [unknown, boolean, string];
}

export interface Middleware {
  execute(
    toolName: string,
    args: Record<string, unknown>,
    caller: CallerContext
  ): [unknown, boolean, string];
  withPolicy(evaluator: PolicyEvaluator): MiddlewareImpl;
  withAuditor(auditor: Auditor): MiddlewareImpl;
}

export class MiddlewareImpl implements Middleware {
  private executor: ToolExecutor;
  private evaluator: PolicyEvaluator | null = null;
  private auditor: Auditor | null = null;

  constructor(executor: ToolExecutor) {
    this.executor = executor;
  }

  execute(
    toolName: string,
    args: Record<string, unknown>,
    caller: CallerContext
  ): [unknown, boolean, string] {
    const decision = this.evaluator
      ? this.evaluator.evaluate(toolName, args, caller)
      : { action: Action.ALLOW, rule: "default", reason: "no policy configured", timestamp: new Date() };

    if (this.auditor) {
      this.auditor.record({
        session_id: caller.session_id || "",
        tool_name: toolName,
        args,
        decision,
        timestamp: new Date(),
      });
    }

    if (decision.action === Action.DENY) {
      return [null, false, `denied by policy: ${decision.reason}`];
    }

    if (decision.action === Action.FILTER && decision.redacted_args) {
      args = { ...args, ...decision.redacted_args };
    }

    return this.executor.execute(toolName, args);
  }

  withPolicy(evaluator: PolicyEvaluator): MiddlewareImpl {
    this.evaluator = evaluator;
    return this;
  }

  withAuditor(auditor: Auditor): MiddlewareImpl {
    this.auditor = auditor;
    return this;
  }
}

export function newMiddleware(executor: ToolExecutor): MiddlewareImpl {
  return new MiddlewareImpl(executor);
}

File: go/middleware/middleware.go
package middleware

import (
	"context"

	"github.com/soypete/pedro-agentware/go/tools"
)

type ToolExecutor interface {
	Execute(ctx context.Context, toolName string, args map[string]any) (*tools.Result, error)
}

type Middleware interface {
	ToolExecutor
	WithPolicy(evaluator PolicyEvaluator) Middleware
	WithAuditor(auditor Auditor) Middleware
}

type middlewareImpl struct {
	exec      ToolExecutor
	evaluator PolicyEvaluator
	auditor   Auditor
}

func NewMiddleware(exec ToolExecutor) Middleware {
	return &middlewareImpl{
		exec: exec,
	}
}

func (m *middlewareImpl) Execute(ctx context.Context, toolName string, args map[string]any) (*tools.Result, error) {
	caller := getCallerContext(ctx)

	var decision Decision
	if m.evaluator != nil {
		decision = m.evaluator.Evaluate(toolName, args, caller)
	} else {
		decision = Decision{Action: ActionAllow, Reason: "no policy configured"}
	}

	auditRecord := AuditRecord{
		SessionID: caller.SessionID,
		ToolName:  toolName,
		Args:      args,
		Decision:  decision,
	}
	if m.auditor != nil {
		m.auditor.Record(auditRecord)
	}

	if decision.Action == ActionDeny {
		return &tools.Result{
			Success: false,
			Error:   "denied by policy: " + decision.Reason,
		}, nil
	}

	if decision.Action == ActionFilter && len(decision.RedactedArgs) > 0 {
		for k, v := range decision.RedactedArgs {
			args[k] = v
		}
	}

	return m.exec.Execute(ctx, toolName, args)
}

func (m *middlewareImpl) WithPolicy(evaluator PolicyEvaluator) Middleware {
	m.evaluator = evaluator
	return m
}

func (m *middlewareImpl) WithAuditor(auditor Auditor) Middleware {
	m.auditor = auditor
	return m
}

func getCallerContext(ctx context.Context) CallerContext {
	if c, ok := ctx.Value(callerContextKey).(CallerContext); ok {
		return c
	}
	return CallerContext{
		Trusted: true,
	}
}

type contextKey string

const callerContextKey contextKey = "caller_context"

func WithCallerContext(ctx context.Context, caller CallerContext) context.Context {
	return context.WithValue(ctx, callerContextKey, caller)
}
